### PR TITLE
Frank!Doc - Add monitoring to the Frank!Doc

### PR DESCRIPTION
**Pull request by mhdirkse**
Supersedes https://github.com/ibissource/iaf/pull/1930.

I did the following tests:
- Compare the monitors shown in the Frank!Console with the same view from the master branch.
- Try the "to XML" buttons for the monitors and compare the resulting XMLs with original XML of the configuration.
- Create an example monitor and check that it can be triggered. I had an example monitor that logs to the GALM log. I triggered the monitor using Test a Pipeline. I checked that the monitor was raised in the GALM and in the Frank!Console. Then I cleared the monitor successfully using the Frank!Console.
- Run the Larva tests.
- In the example configuration file MonitorConfig.xml in ibis-adapterframework-test, change each element name to a capital. Check that the result satisfies the Frank!Doc strict XSD.
- Check that monitoring appears correctly in the Frank!Doc website.

Known issues:
- Editing monitors does not work properly. You get the dialogs to do the editing, but the monitors do not behave as expected afterwards.
- The Frank!Doc strict XSD does not accept elements that are all lower-case. The first letter must be a capital. But if you update the example file MonitorConfig.xml in ibis-adapterframework-test, then the monitors are not shown in the Frank!Console anymore.

### CODE REVIEW

**Comment by mhdirkse** [comment]
Please merge https://github.com/ibissource/iaf/pull/2136 first. After this merge, I will merge the master to this pull request manually.

**Comment by mhdirkse** [comment]
I have tested this PR again. I did the following:
* I checked that the Frank!Doc website showed Alarm, Clearing and Trigger in the right way.
* I compared the console page with monitors with the contents you have with master. These are the monitors of ibis4test.
* Also in ibis4test, I checked the XML you get from the toXML buttons of the monitors. I compared those XML strings with the original XML that defined the monitors.
* I fired and cleared a test monitor I hacked into ibis4example. I did so both by running an adapter and by raising/clearing my test monitor in the console.

While testing, I found issue https://github.com/ibissource/iaf/issues/2162. That issue is not solved in this PR.

**Comment by mhdirkse** [comment]
I do not make this PR ready-for-review only because of the sequence of merging. Please merge PR https://github.com/ibissource/iaf/pull/2161 first.

File: core/src/main/java/nl/nn/adapterframework/monitoring/MonitorManager.java [hunk 1]
```diff
@@ -199,23 +201,19 @@ public XmlBuilder getStatusXml() {
...
+			configXml.addSubElement(ma.toXml());
 		}
-		configXml.addSubElement(destinationsXml);
-		XmlBuilder monitorsXml=new XmlBuilder("monitors");
 		for (int i=0; i<monitors.size(); i++) {
 			Monitor monitor=getMonitor(i);
-			monitorsXml.addSubElement(monitor.toXml());
+			configXml.addSubElement(monitor.toXml());
 		}
-		configXml.addSubElement(monitorsXml);
```
Comment (gvanbrakel) [inline core/src/main/java/nl/nn/adapterframework/monitoring/MonitorManager.java:218]: Hier zijn nu de container elementen 'destinations' en 'monitors' weggehaald. Daar wordt de structuur van het XmlDocument niet beter op.

File: test/src/main/configurations/TX/MonitorConfig.xml [hunk 2]
```diff
@@ -1,76 +1,51 @@
...
+			<adapterfilter adapter="TransactionHandlingMandatory">
+				<source>Pipe [Send transacted message to mandatory subadapter]</source>
+			</adapterfilter>
+			<adapterfilter adapter="TransactionHandlingMandatorySubAdapter">
+				<source>Pipe [Continue if this call was transacted]</source>
+			</adapterfilter>
+		</trigger>
+	</monitor>
+	<monitor name="Sender exception" type="TECHNICAL" destinations="GALM">
+		<trigger className="nl.nn.adapterframework.monitoring.Alarm" eventCode="Sender Exception Caught" severity="CRITICAL" threshold="10" period="3600"/>
```
Comment (gvanbrakel) [inline test/src/main/configurations/TX/MonitorConfig.xml:23]: In plaats van 
`<trigger className="nl.nn.adapterframework.monitoring.Alarm" ... />`
verwacht ik
`<AlarmTrigger ... />`

File: test/src/main/configurations/TX/MonitorConfig.xml [hunk 3]
```diff
@@ -1,76 +1,51 @@
...
+		<trigger className="nl.nn.adapterframework.monitoring.Alarm" eventCode="Exception Configuring Receiver" severity="CRITICAL"/>
+	</monitor>
+	<monitor name="Message size exceeding" type="TECHNICAL">
+		<trigger className="nl.nn.adapterframework.monitoring.Alarm" eventCode="Pipe Message Size Exceeding" severity="WARNING"/>
+	</monitor>
+	<monitor name="Message in Error" type="FUNCTIONAL" destinations="GALM">
+		<trigger className="nl.nn.adapterframework.monitoring.Alarm" severity="WARNING">
+			<event>Receiver Moved Message to ErrorStorage</event>
+		</trigger>
+		<trigger className="nl.nn.adapterframework.monitoring.Clearing" severity="WARNING">
```
Comment (gvanbrakel) [inline test/src/main/configurations/TX/MonitorConfig.xml:39]: In plaats van
<trigger className="nl.nn.adapterframework.monitoring.Clearing" ... />
verwacht ik
<ClearingTrigger ... />

**Review by gvanbrakel — CHANGES_REQUESTED** [review]

File: core/src/main/java/nl/nn/adapterframework/util/Misc.java [hunk 4]
```diff
@@ -1413,4 +1413,8 @@ public static String jsonPretty(String json) {
 
 		return sw.toString().trim();
 	}
+
+	public static String toUpperCamelCase(String arg) {
+		return arg.substring(0,  1).toUpperCase() + arg.substring(1);
+	}
```
Comment (nielsm5) [inline core/src/main/java/nl/nn/adapterframework/util/Misc.java:1419]: Dit is initialCaps, en kan nu niet overweg met upper case strings

File: core/src/main/java/nl/nn/adapterframework/monitoring/Trigger.java [hunk 5]
```diff
@@ -192,32 +194,32 @@ public Monitor getMonitor() {
...
-	public String getType() {
-		if (isAlarm()) {
-			return "Alarm";
-		} else {
-			return "Clearing";
-		}
+	public String getTypeDisplayString() {
+		String nameLC = triggerType.name().toLowerCase();
+		return Misc.toUpperCamelCase(nameLC);
+	}
```
Comment (nielsm5) [inline core/src/main/java/nl/nn/adapterframework/monitoring/Trigger.java:214]: Deze methode hoort hier niet.

File: core/src/main/java/nl/nn/adapterframework/webcontrol/api/ShowMonitors.java [hunk 6]
```diff
@@ -163,7 +163,7 @@ public Response getMonitors(@PathParam("configuration") String configurationName
 	private Map<String, Object> mapTrigger(ITrigger trigger) {
 		Map<String, Object> triggerMap = new HashMap<String, Object>();
 
-		triggerMap.put("type", trigger.getType());
+		triggerMap.put("type", trigger.getTypeDisplayString());
```
Comment (nielsm5) [inline core/src/main/java/nl/nn/adapterframework/webcontrol/api/ShowMonitors.java:166]: Hier moet hij misc.initialCaps(enum.name()) kunnen doen. Met de extra toLowerCase in de initialCaps methode

**Review by nielsm5 — CHANGES_REQUESTED** [review]

**Comment by nielsm5** [comment]
Op mijn laatste puntje na is het goed!

File: core/src/main/java/nl/nn/adapterframework/webcontrol/api/ShowMonitors.java (hunk 6 above)
Comment (mhdirkse) [reply core/src/main/java/nl/nn/adapterframework/webcontrol/api/ShowMonitors.java]: Hopelijk gedaan. Is dit wat je bedoelt?

File: core/src/main/java/nl/nn/adapterframework/monitoring/Trigger.java (hunk 5 above)
Comment (mhdirkse) [reply core/src/main/java/nl/nn/adapterframework/monitoring/Trigger.java]: Hopelijk gedaan. Is dit wat je bedoelt?

File: core/src/main/java/nl/nn/adapterframework/util/Misc.java (hunk 4 above)
Comment (mhdirkse) [reply core/src/main/java/nl/nn/adapterframework/util/Misc.java:1419]: Ik wilde toUpperCamelCase niet veranderen. Die gebruik ik veel in de Frank!Doc. Is het zo OK?

**Review by gvanbrakel — APPROVED** [review]

**Review by nielsm5 — APPROVED** [review]

File: frankDoc/src/main/java/nl/nn/adapterframework/frankdoc/Utils.java [hunk 7]
```diff
@@ -101,6 +100,6 @@ public static String promoteIfPrimitive(String typeName) {
 	}
 
 	public static String toUpperCamelCase(String arg) {
-		return Misc.toUpperCamelCase(arg);
+		return arg.substring(0,  1).toUpperCase() + arg.substring(1);
```
Comment (gvanbrakel) [inline frankDoc/src/main/java/nl/nn/adapterframework/frankdoc/Utils.java]: Dit is geen upperCamelCase, maar initialCaps

**Review by gvanbrakel — APPROVED** [review]

**Pull request merged**